cmake/modules/RootMacros.cmake

```diff
@@ -283,115 +283,226 @@ function(ROOT_GENERATE_DICTIONARY dictionary)
     set(libprefix "")
   endif()
 
-  #---Get the list of include directories------------------
-  get_directory_property(incdirs INCLUDE_DIRECTORIES)
-  # rootcling invoked on foo.h should find foo.h in the current source dir,
-  # no matter what.
-  list(APPEND incdirs ${CMAKE_CURRENT_SOURCE_DIR})
+  if((CMAKE_PROJECT_NAME STREQUAL ROOT) AND (TARGET ${ARG_MODULE}))
+    set(incdirs)
+
+    if (cxxmodules OR runtime_cxxmodules)
+       # Comments from Vassil:
+       # FIXME: We prepend ROOTSYS/include because if we have built a module
+       # and try to resolve the 'same' header from a different location we will
+       # get a redefinition error.
+       # We should remove these lines when the fallback include is removed. Then
+       # we will need a module.modulemap file per `inc` directory.
+       list(APPEND incdirs ${CMAKE_BINARY_DIR}/include)
+    else()
+       # Comments from Sergey: because of TVirtualObject.h, appearing in each dictionary
+       list(APPEND incdirs ${CMAKE_SOURCE_DIR}/io/io/inc)
+    endif()
+
+    set(headerdirs)
 
-  if(TARGET ${ARG_MODULE})
     get_target_property(target_incdirs ${ARG_MODULE} INCLUDE_DIRECTORIES)
-    foreach(dir ${target_incdirs})
-      string(REGEX REPLACE "^[$]<BUILD_INTERFACE:(.+)>" "\\1" dir ${dir})
-      if(NOT ${dir} MATCHES "^[$]")
-        list(APPEND incdirs ${dir})
-      endif()
+    if(target_incdirs)
+       foreach(dir ${target_incdirs})
+          string(REGEX REPLACE "^[$]<BUILD_INTERFACE:(.+)>" "\\1" dir ${dir})
+          if(NOT ${dir} MATCHES "^[$]")
+             list(APPEND incdirs ${dir})
+          endif()
+          if(${dir} MATCHES "^${CMAKE_SOURCE_DIR}")
+             list(APPEND headerdirs ${dir})
+          endif()
+       endforeach()
+    endif()
+
+
+    # Comments from Sergey
+    # FIXME: we have to exclude all source directories from includes while they duplicated in ${CMAKE_BINARY_DIR}/include
+    # rootcling is not able to handle such situation, therefore ${CMAKE_BINARY_DIR}/include is used as source for ROOT headers
+    # Also ${CMAKE_BINARY_DIR}/ginclude is removed
+    # add slash at the end to prevent filtering include paths from roottest which may have same location as root
+    list(FILTER incdirs EXCLUDE REGEX "^${CMAKE_SOURCE_DIR}\/")
+    list(FILTER incdirs EXCLUDE REGEX "^${CMAKE_BINARY_DIR}/ginclude")
+    list(FILTER incdirs EXCLUDE REGEX "^${CMAKE_BINARY_DIR}/externals")
+    list(INSERT incdirs 0 ${CMAKE_BINARY_DIR}/include)
+    # end workaround
+
+    if(ARG_MODULE STREQUAL Core)
+       list(APPEND incdirs ${CMAKE_SOURCE_DIR}/core) # This is needed because LinkDef.h includes other LinkDef starting from ${CMAKE_SOURCE_DIR}/core
```

Comment:
Can we reduce this complexity by adjusting the Linkdef includers?

Reply:
I can try to solve it differently - but maybe in other PR

Reply:
Thanks. We should not overload ROOT_GENERATE_DICTIONARY with more complexity than necessary as it is already too complicated. We should drop these changes in favor or your new PR.